### PR TITLE
Fancy error messages

### PR DESCRIPTION
**Pull request by zuiderkwast**
* Colors
* Expressions highlighted in their context (#222)
* A performance hack to fix a bottle neck (#231)

### CODE REVIEW

**Comment by zuiderkwast** [comment]
Btw, for `{color, auto}` I'm using `io:columns()` to detect if stdout is a TTY. It doesn't work from an escript, where it always `{error, enotsup}`. Is there any other way?

One way would be to start the CLI by calling `erl` from a shell script instead of an escript.

File: src/gradualizer_hilite.erl [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+%% Highlights a node in the AST in a piece of source code. The node must
+%% annotated with line and column information for this to work. Returns
+%% an inlist().
+highlight_in_source(AstNode, Source, Color) ->
+    {Start, End} = find_start_and_end_location_in_source(AstNode, Source),
+    highlight_text(Source, Start, End, Color).
+
+%% Finds the start and end location of an abstract node in source code.
+%% This is done by pretty-printing, tokenizing, counting the tokens and
+%% then performing finding the corresponding tokens in the tokenized
```
Comment (Zalastax) [inline src/gradualizer_hilite.erl]: then finding (remove performing)

File: test/gradualizer_hilite_tests.erl [hunk 2]
```diff
@@ -0,0 +1,20 @@
+-module(gradualizer_hilite_tests).
+
```
Comment (Zalastax) [inline test/gradualizer_hilite_tests.erl]: gradualizer_highlight_tests?

File: src/gradualizer_hilite.erl [hunk 3]
```diff
@@ -0,0 +1,233 @@
+-module(gradualizer_hilite).
+
```
[the remark this answers is not in the document]
Comment (zuiderkwast) [reply src/gradualizer_hilite.erl]: it's too long :-) but I can change if you want

**Comment by Zalastax** [comment]
I think some discussion is needed about the performance hack. Do we have problems with all unions? Is our approach too naive? How problematic is it to replace with any?

**Comment by zuiderkwast** [comment]
> I think some discussion is needed about the performance hack. Do we have problems with all unions? Is our approach too naive? How problematic is it to replace with any?

I agree, but I would like to merge it first and discuss afterwards. Running out of memory is not OK. Let's discuss and/or find a proper solution in #231.

**Comment by gomoripeti** [comment]
awesome work!

(I tried to understand how rebar3 escript does tty detection. I kindof remembered escripts use the old-shell. I couldn't find any auto-color detection snippet. I only found that `rebar3 shell` basically stops the old-shell and starts a new-type shell https://github.com/erlang/rebar3/blob/master/src/rebar_prv_shell.erl#L140 - the horror)

**Comment by gomoripeti** [comment]
Just bumped into this erlang issue ticket about `io:columns` in escript fixed in OTP 22.1. Not too relevant but interesting https://bugs.erlang.org/browse/ERL-717

**Comment by zuiderkwast** [comment]
> 
> 
> Just bumped into this erlang issue ticket about `io:columns` in escript fixed in OTP 22.1. Not too relevant but interesting https://bugs.erlang.org/browse/ERL-717

That's great! It means that automatic colors will work.

File: src/gradualizer_fmt.erl [hunk 4]
```diff
@@ -371,20 +373,64 @@ format_type_error({bad_type_annotation, TypeLit}, Opts) ->
...
+             {ok, SourceBin} = file:read_file(SourceFile),
+             Source = binary_to_list(SourceBin),
+             gradualizer_highlight:highlight_in_source(AstNode, Source, Color)
+     end,
+     $\n]. % blank line after
+
+-spec use_color(gradualizer:options()) -> boolean().
+use_color(Opts) ->
+    case proplists:get_value(color, Opts, auto) of
+        auto -> case io:columns() of {ok, _} -> true; _ -> false end;
```
Comment (josefs) [inline src/gradualizer_fmt.erl]: Can you explain the use of `io:columns/0` here? It's be nice to have a comment in the code detailing what's going on here.

**Pull request merged**